momepy/preprocessing.py

```diff
@@ -740,3 +746,331 @@ def _get_extrapolated_line(coords, tolerance, point=False):
+    tmp["geometry"] = tmp.geometry.values.data
+
+    pygeos_geoms = (
+        tmp.groupby("index_right")
+        .geometry.apply(pygeos.multipolygons)
+        .rename("geometry")
+    )
+    pygeos_geoms = pygeos.make_valid(pygeos_geoms)
+
+    rab_multipolygons = gpd.GeoDataFrame(pygeos_geoms, crs=gdf.crs)
+    # make_valid is transforming the multipolygons into geometry collections because of
+    # shared edges
+
+    if center_type == "centroid":
+        # geometry centroid of the actual circle
+        rab_multipolygons["center_pt"] = gdf[
+            gdf.index == gdf.index_right
+        ].geometry.centroid
+
+    elif center_type == "mean":
+        coords, idxs = pygeos.get_coordinates(pygeos_geoms, return_index=True)
+        means = {}
+        for i in np.unique(idxs):
+            tmps = coords[idxs == i]
+            target_idx = rab_multipolygons.index[i]
+            means[target_idx] = Point(tmps.mean(axis=0))
+
+        rab_multipolygons["center_pt"] = gpd.GeoSeries(means, crs=gdf.crs)
+
+    # centerpoint of minimum_bounding_circle
+    # minimun_bounding_circle() should be available in Shapely 2.0. Implementation still
+    # pending.
+    # current environment has 1.8.2
+
+    return rab_multipolygons
+
+
+def _coins_filtering_many_incoming(incoming_many, angle_threshold=0):
+    """
+    Used only for the cases when more than one incoming line touches the
+    roundabout.
+    """
+    coins_filter_result = []
+    # For each new connection, evaluate COINS and select the group from which the new
+    # line belongs
+    # TODO ideally use the groupby object on line_wkt used earlier
+    for g, x in incoming_many.groupby("line_wkt"):
+        gs = gpd.GeoSeries(pd.concat([x.geometry, x.line]), crs=incoming_many.crs)
+        gdf = gpd.GeoDataFrame(geometry=gs)
+        gdf.drop_duplicates(inplace=True)
+
+        coins = COINS(gdf, angle_threshold=angle_threshold)
+        # coins.stroke_attribute()) # the groups here don't match the stroke_group in
+        # .stroke_gdf()
+        stroke_gdf = coins.stroke_gdf()
+        if GPD_10:
+            orig_geom_join = stroke_gdf.sjoin(
+                gpd.GeoDataFrame(geometry=x.line), predicate="covers"
+            )
+        else:
+            orig_geom_join = stroke_gdf.sjoin(
+                gpd.GeoDataFrame(geometry=x.line), op="covers"
+            )
+        orig_geom = gpd.GeoSeries(
+            [orig_geom_join.geometry.iloc[0]], crs=incoming_many.crs
+        )
+        gs2 = gpd.GeoDataFrame(geometry=orig_geom)
+
+        gs1 = gpd.GeoSeries(x.geometry, crs=incoming_many.crs)
+        gs1 = gpd.GeoDataFrame(geometry=gs1)
+
+        # select the the line that's covered by the joined line returned by COINS
+        # one could consider using pygeos shared_paths(a, b) # TODO
+        if GPD_10:
+            result_idx = gs1.sjoin(gs2, predicate="covered_by").index
+        else:
+            result_idx = gs1.sjoin(gs2, op="covered_by").index
```

Comment:
```suggestion
            result_idx = gpd.sjoin(gs1, gs2, op="covered_by").index
```

Same as before.

You could replace the sjoin here with an `sindex` query though, to avoid the overhead of sjoin but I'd need to run this to actually write that snippet, so feel free to ignore this comment.

Reply:
not sure I fully understand what you mean... but would be curious to see what you suggest to still maintain the list of indexes as an output of the  function (which is later used to remove the original lines from the edges gdf)

Reply:
Take this example:

```py
In [1]: import geopandas
world = geopa                                                                  
In [2]: world = geopandas.read_file(geopandas.datasets.get_path('naturalearth_lowres'))

In [3]: cities = geopandas.read_file(geopandas.datasets.get_path('naturalearth_cities'))
```

This is your way - do the whole sjoin and then extract only the index to know what should be dropped.

```py
In [4]: sjoin_idx = cities.sjoin(world).index

In [5]: sjoin_idx
Out[5]: 
Int64Index([  0,   1, 192,   2, 184,   3,   8,  11, 199,  12,
            ...
            188, 190, 191, 193, 194, 195, 196, 197, 198, 200],
           dtype='int64', length=172)

```
Under the hood, geopandas calls `.sindex.query_bulk` to know which geometries intersect each other (or are covered by if `predicate="covered_by"`) and does a lot of other things to construct the final GeoDataFrame. But we're not interested in those other things.

So you can call just the sindex yourself:

```py
In [6]: world_idx, cities_idx = cities.sindex.query_bulk(world.geometry, predicate="intersects")
```
We get two arrays out, representing pairs of geometries in each gdf that fulfil the predicate. Note, that these are integer (positional) indices. 

See that the one for `cities` has the same length as yours and since `cities.index` is a default RangeIndex, hence its values are the same as a position, all values actually match, they are just sorted in a different order. 
```

In [7]: len(cities_idx)
Out[7]: 172

In [8]: all([x in sjoin_idx for x in cities_idx])
Out[8]: True
```

If the original index wasn't the default one, you get the result you want like this:

```py
In [9]: cities.index.take(cities_idx)
Out[9]: 
Int64Index([ 91, 144, 101, 186,  96, 165,  59, 195, 182, 200,
            ...
            123,  48,  13, 120,  25, 102,  19,  21,  12,  15],
           dtype='int64', length=172)
```

The reason to do this is to avoid the overhead of sjoin. This is not the best test case, but you can check the difference in performance yourself.

```py
In [10]: %timeit cities.sjoin(world).index
4.08 ms ± 11.3 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

In [11]: %timeit cities.sindex.query_bulk(world.geometry, predicate="intersects")
425 µs ± 482 ns per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
```

Reply:
I just included this but `query_bulk` doesn't have the predicate `'covered_by'` so I ended up changing some things to still get the desired output...

```
`edges_idx,` rabs_idx = rab_multipolygons.buffer(1).sindex.query_bulk(
            edges.geometry, predicate="within")
```

 I couldn't find a way to use any of the predicates without using the buffer

Reply:
It does have all the predicates. Without that we couldn't expose it in sjoin. See https://pygeos.readthedocs.io/en/stable/strtree.html#pygeos.strtree.STRtree.query_bulk

Reply:
of course, that makes sense since it's under the hood of spatial joins... 
I was looking here instead: https://geopandas.org/en/stable/docs/reference/api/geopandas.sindex.SpatialIndex.query_bulk.html

This is what I keep testing and still get different results than doing the original `sjoin`

> rabs_idx, edges_idx = edges.sindex.query_bulk(rab_multipolygons.geometry, predicate= "covered_by")
> idx_drop = edges.index.take(edges_idx)
-> `idx_drop`  is empty in this case

> idx_drop = edges.sjoin(rab_multipolygons, predicate="covered_by").index
-> `idx_drop` actually has the roundabout lines (excluding the incoming lines)

Any idea what could be happening? (I feel it has to be an oversight from my side)

Reply:
It is not the most intuitive but the sindex query is working in an opposite direction to sjoin

- sindex: return _edges_ that are _"covered_by"_ by _rab_multipolygons.geometry_
- sjoin: join _rab_multipolygons_ to _edges_ if _rab_multipolygons_ are _"covered by"_ by _edges_

So the query needs to be opposite.